lib/src/utils.dart

```diff
@@ -29,3 +29,64 @@ void validateErrorArgs(
         'the string.');
   }
 }
+
+// See https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF
+// for documentation on how UTF-16 encoding works and definitions of various
+// related terms.
+
+/// The lower bound of Unicode's supplementary plane.
+const _supplementaryPlaneLowerBound = 0x10000;
+
+/// The lower bound of the UTF-16 high surrogate block.
+const _highSurrogateLowerBound = 0xD800;
+
+/// The lower bound of the UTF-16 low surrogate block.
+const _lowSurrogateLowerBound = 0xDC00;
+
+/// The number of low bits in each code unit of a surrogate pair that goes into
+/// determining which code point it encodes.
+const _surrogateBits = 10;
+
+/// A bit mask that covers the lower bits of a code point, which are then
+/// encoded into its low surrogate.
+const _lowSurrogateMask = (1 << _surrogateBits) - 1;
+
+/// Returns whether [codePoint] is in the Unicode supplementary plane, and thus
+/// must be represented as a surrogate pair in UTF-16.
+bool inSupplementaryPlane(int codePoint) =>
+    codePoint >= _supplementaryPlaneLowerBound;
```

Comment:
Should have an upper bound too, of 0x10FFFF.
Otherwise `highSurrogate` below can overflow, since it doesn't mask off only ten bits.

I guess the test should actually be somewhere else, because we don't want too big numbers to fall into the BMP branch either. So maybe validate user character input, which I guess is primarily `scanChar`, with:
```dart
RangeError.checkValueInInterval(character, 0, 0x10FFFF, "character");
```

The test for whether this works could be:
```dart
var s = "\uD800\uE000";
var c = StringScanner(s);
c.scanChar(0x110000); // Should not return true.
```

Reply:
I've added tests to ensure that this behaves correctly, but I don't think it's worth adding a range check to `scanChar()`. That function is often called in tight loops, and I think there's value in only having it return `true` or `false` and never erroring out.